### ISSUE TITLE
Use a better byte buffer

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects/spring-reactive#7]
`java.nio.ByteBuffer` is a great basis but it suffers from several limitations:
 - No auto-expand
 - Need to use `flip()` when switching between reading and writing
 - Need real allocation management, maybe with reference counting
 - Need additional functionalities like combining easily 2 `ByteBuffer` without copying the data.

Currently we are using Reactor `Buffer` internally but we may eventually create and use a better implementation. Netty `ByteBuf` is not an option essentially because this byte buffer type may be exposed in our API. The better byte buffer is likely to use or extends `java.nio.ByteBuffer`.

Some interesting reading about that topic:
 - http://www.evanjones.ca/software/java-bytebuffers.html
 - http://flink.apache.org/news/2015/09/16/off-heap-memory.html
 - [XNIO new ByteBufferPool](https://github.com/xnio/xnio/blob/3.x/api/src/main/java/org/xnio/ByteBufferPool.java)
 - [Reactor Buffer](https://github.com/reactor/reactor/blob/master/reactor-core/src/main/java/reactor/io/buffer/Buffer.java) (we discussed with @smaldini to extends ByteBuffer instead of wrapping it)
 - [Agrona buffer interfaces and implementation](http://insightfullogic.com/2015/Apr/18/agronas-threadsafe-offheap-buffers/) ([DirectBuffer](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/DirectBuffer.java), [MutableDirectBuffer](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/MutableDirectBuffer.java) and [UnsafeBuffer](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/concurrent/UnsafeBuffer.java)) are also worth to see (@smaldini and @thomasdarimont both mentioned that project)

**Comment by rstoyanchev** [comment]
These are some of the advantages of the (Netty) ByteBuf API:
- It is extensible to user-defined buffer types.
- Transparent zero-copy is achieved by a built-in composite buffer type.
- Capacity is expanded on demand, as with the JDK StringBuilder.
- There is no need to call flip() to switch between reader and writer modes.
- Reader and writer employ distinct indices.
- Supports method chaining.
- Supports reference counting.
- Supports pooling.

**Comment by bclozel** [comment]
Note: @smaldini is also looking at Atomic Buffers for use in concurrent environments.

See [Agrona AtomicBuffer](https://github.com/real-logic/Agrona/blob/master/src/test/java/uk/co/real_logic/agrona/concurrent/AtomicBufferTest.java).

**Comment by sdeleuze** [comment]
I have tried to summarize what are my main expectations for the `Buffer` type:
 - Support reading/writing concurrently data in different threads
 - `Java.nio.ByteBuffer` API is not designed for that so we don't want to extend it or expose its methods directly
 - No flip() method to switch between reading/writing
 - We should have easy `ByteBuffer` interoperability
 - Auto-expand capabilities
 - Combining N Buffers into a single virtual one with 0 copy mechanism
 - Supports reference counting
 - Supports pooling, eventually including off-heap
 - Primitive encoding/decoding (including `String` with optionally specified charset)

**Comment by poutsma** [comment]
Let's summarize the options when it comes to byte buffer usage in Spring:

1. Use `java.nio.ByteBuffer`. Upside: we use the standard JDK mechanism. Downside: as @sdeleuze has already indicated above, JDK's `ByteBuffer` sucks.
2. Use an existing 3rd party byte buffer, such as Netty's `ByteBuf`. Upside: we don't have to do any work. Downside: this would add and expose a required dependency, which is highly unusual of Spring.
3. Create an abstraction over existing (3rd party) implementations, detecting at runtime which is available and possibly falling back to the JDK `ByteBuffer`. Upside: we don't have to do a lot of work, only the abstraction. Downside: since it is an abstraction, API usage will become more cumbersome (i.e. creating a buffer via a factory instead of constructing one directly). This option also still requires 3rd party dependencies for optimal performance, we're just hiding it behind an interface.
4. Create our own buffer type, somewhere in `org.springframework.core`. Upside: we are in direct control of the code, and can tune it to our use cases. Downside: writing (and maintaining !) buffer code can be quite complex, though there are many sources of inspiration out there.

My personal preference at this time is option 4, as it has the best upside/downside balance.

**Comment by poutsma** [comment]
Goals for byte buffer, as decided on 2015110204 meeting:

1. Avoid required 3rd party dependency
2. Easier to use than ByteBuffer
3. Efficient Network I/O (i.e. consider pooling, reference counting, zero-copy)
4. Capable of dealing with concurrency

**Comment by sdeleuze** [comment]
Maybe I could refactor a little bit our codebase to use a small Spring Reactive byte buffer interface instead of exposing directly `java.nio.ByteBuffer` with only the methods we are using for now (except `flip()`I don't want to expose). That way we could experiment easily with different implementation.

Even if our final goal is to avoid requiring 3rd party dependencies, we could then create branches with Agrona buffer based implementation and Netty `ByteBuf` implementation to better understand the pros/cons of these implementation more concretely. That could be a nice first step before implementing our own.

Any thoughts?

**Comment by sdeleuze** [comment]
After different tries to find something simple, meaningful and different from JDK `ByteBuffer` Reactor `Buffer` and Netty `ByteBuf`, I suggest `Bytes` for this (maybe not final but at least temporary until we found something better) Spring Reactor byte buffer interface name.

**Comment by bclozel** [comment]
I like `Bytes`; very simple, and easy to read and pronounce. Also makes sense when reading/writing `Publisher<Bytes>`. There's already the `java.lang.Byte` class, but we're still pretty high in the autocompletion list.

![image](https://cloud.githubusercontent.com/assets/103264/10973748/00b42f9a-83df-11e5-91e5-c4fa1dad8c3c.png)

Now another related question: should static builder methods be included in that class or in a separate Utility class?

**Comment by smaldini** [comment]
Bytes is not bad at all indeed. It's transparent on the type of information carried and Buffer is actually an oxymora for an object holding bytes!

**Comment by poutsma** [comment]
@sdeleuze: `Bytes` to me sounds like the utility companion class for `Byte`. As in: `java.util.Collection` vs `java.util.Collections`, `java.lang.Object` vs `java.util.Objects`, etc. This is quite a common pattern in the JDK (though not in Spring), so perhaps we should stay clear of it.

@bclozel: The fact that `Bytes` appears high in the autocompletion list is not really that relevant, as we agreed that it’s going to be an internal API. So any autocompletion niceness is mostly going to benefit us as Spring developers, which isn't that important.

@smaldini: I don't think buffer is a oxymoron (or pleonasm, which I think you meant ?) for an object holding bytes. There are other usages of the term 'buffer' in software: char buffer, frame buffer, etc. 

Is there anything wrong with calling it a `ByteBuffer`, like everybody else does? As it's an internal API, name clashes would mostly only bother us, not typical end users.

**Comment by bclozel** [comment]
While I agree that this type shouldn't be "mainstream" in typical applications, this type is still an important part of our request/response contract.
So it should be user-facing when writing raw handlers or a `RequestInterceptor`, right?

**Comment by poutsma** [comment]
Indeed, and that's why I used the word 'mostly'. I would argue that the vast majority of our users doesn't write interceptors.

**Comment by sdeleuze** [comment]
I agree that `ByteBuffer` is the most meaningful class name, but I think using it will be **very** confusing since `java.nio.ByteBuffer` is widely used and known. Even in our discussion we will be confused, because when talking about `ByteBuffer` we won't know if we talk about the `java.nio` one or about the Spring one.

Also I am not sure to see what you mean by "like everybody else does". Netty uses `ByteBuf`, Reactor use `Buffer`, Agrona use other class names than `ByteBuffer` to avoid this kind of confusion.

This type will not only be exposed in interceptors, but also in Reactive(Server|Client)HttpRequest#(get/set)Body. This will be also be supported an handler method input parameter or return value (`Publisher<...>`). So this is clearly part of our public API.

**Comment by poutsma** [comment]
> I agree that ByteBuffer is the most meaningful class name, but I think using it will be very confusing since java.nio.ByteBuffer is widely used and known. Even in our discussion we will be confused, because when talking about ByteBuffer we won't know if we talk about the java.nio one or about the Spring one.

That's true, and a good reason not to use `ByteBuffer`.

> 
> Also I am not sure to see what you mean by "like everybody else does". Netty uses ByteBuf, Reactor use Buffer, Agrona use other class names than ByteBuffer to avoid this kind of confusion.

I was referring to this:




Turns out that a lot of frameworks use ByteBuffer as a class name (but not Netty, Reactor nor Agrona). Though in fairness, there are some duplicates in there.

> This type will not only be exposed in interceptors, but also in Reactive(Server|Client)HttpRequest#(get/set)Body. This will be also be supported an handler method input parameter or return value (Publisher<...>). So this is clearly part of our public API.

It is part of our public API, but I would argue that the majority of our users will not deal with the body contents as a byte buffer. Just like currently, in @MVC, they do not deal with the `ServerHttpRequest.getBody()` input stream. They will use @RequestBody/@ResponseBody or similar constructs.

Overall, I guess we can use `Bytes` for now, but I do hope we come up with a better name for it later.

**Comment by poutsma** [comment]
> I agree that ByteBuffer is the most meaningful class name, but I think using it will be very confusing since java.nio.ByteBuffer is widely used and known. Even in our discussion we will be confused, because when talking about ByteBuffer we won't know if we talk about the java.nio one or about the Spring one.

That's true, and a good reason not to use `ByteBuffer`.

> 
> Also I am not sure to see what you mean by "like everybody else does". Netty uses ByteBuf, Reactor use Buffer, Agrona use other class names than ByteBuffer to avoid this kind of confusion.

I was referring to this:

<img width="1101" alt="screen shot 2015-11-06 at 11 31 18" src="https://cloud.githubusercontent.com/assets/330665/10995382/a1ae7b24-847c-11e5-9b29-8dc498e525cb.png">


Turns out that a lot of frameworks use ByteBuffer as a class name (but not Netty, Reactor nor Agrona). Though in fairness, there are some duplicates in there.

> This type will not only be exposed in interceptors, but also in Reactive(Server|Client)HttpRequest#(get/set)Body. This will be also be supported an handler method input parameter or return value (Publisher<...>). So this is clearly part of our public API.

It is part of our public API, but I would argue that the majority of our users will not deal with the body contents as a byte buffer. Just like currently, in MVC, they do not deal with the `ServerHttpRequest.getBody()` input stream. They will use RequestBody/ResponseBody or similar constructs.

Overall, I guess we can use `Bytes` for now, but I do hope we come up with a better name for it later.

**Comment by sdeleuze** [comment]
I have pushed a refactoring that uses a first `Bytes` interface as part of PR #39. The goal here is just to use exactly the methods we currently use in the codebase, without exposing `java.nio.ByteBuffer` and Reactor `Buffer` types.

This will allow us to experiment about what `Bytes` interface should be, and about various implementation.

**Comment by poutsma** [comment]
I am still not sure about the name `Bytes`. This type is supposed to represent a byte buffer: a region of memory, either on the heap or direct, represented as bytes. The *memory buffer* part is as important as the *bytes* part. So by using `Bytes`, I feel like we're only sending out half of the message we're trying to get across.

So let me throw out a bunch of alternative names out here, perhaps we like one:

* `MemoryStream` (this is what .NET calls its `ByteBuffer`, though in fairness the two worlds are rather different on this front)
 * `MemoryBuffer`
 * `OctectBuffer`
 * `DataBuffer`

**Comment by sdeleuze** [comment]
I think `MemoryStream` will be confusing because the stream aspect is more in Reactive Streams `Publisher` so `Publisher<MemoryStream>` would be confusing, and I am not really fond of `OctectBuffer`.

Maybe `DataBuffer` is more general and a better fit than `MemoryBuffer` (no assumption from where the bytes come from. `DataBuffer` implementation could read the bytes from a file if needed) and since we will provide some methods to read the data as `String`or `Integer` the `Data` term is in fact more accurate than `Byte`!

So I think I would be :+1: for `DataBuffer`!

**Comment by bclozel** [comment]
Same here.
+1 for `DataBuffer`

**Comment by sdeleuze** [comment]
I have also a new one to propose: `ByteChunk`
I think it is my new preferred one ;-)

**Comment by paulvi** [comment]
@poutsma `ByteRegion` ?

**Comment by poutsma** [comment]
Thanks for all the suggestions, after thinking about it a bit more, I think that the buffer suffix is probably more important than the byte prefix. So I am going to stick with `DataBuffer` for now. We can reconsider when we merge with the main Spring codebase.

**Comment by rwinch** [comment]
I intended to send this link as an FYI quite some time ago and it slipped my mind. https://twitter.com/epcjones/status/681221891370041345?s=03

**Comment by poutsma** [comment]
Oh boy.

The tweet seems to be flame bait, just to get more reads. The post itself is much more contextual, showing the specific scenario at *Twitter* that this problem occurred in. This is even admitted by the author in the comments:

>The real world is definitely more nuanced than a tweet allows.

Well, it certainly is a lot more nuanced than this particular tweet allowed for, that much is true!

Another commenter suggests a better tweet/summary:

> Never use Heap Buffer when dealing with MB huge data.

And I can agree to that one.

But within the context of spring reactive, I would argue that we don't do that. For one, our users don't have the extreme use cases that Twitter has. Our typical buffer sizes are not MBs big, but a lot smaller (the default is 256 *bytes*, in both Netty and our own DefaultDataBufferAllocator). And the buffers are mostly allocated by the underlying runtime (eg Netty), not by Spring.

So: nothing to see here, please move along.

**Comment by rstoyanchev** [comment]
Web pages [are obese](http://idlewords.com/talks/website_obesity.htm) these days and we are going to have to get all data at once from template engines today. That said most of that page size comes from images and video so the HTML alone shouldn't be in the MBs indeed.

**Comment by poutsma** [comment]
>Web pages are obese these days and we are going to have to get all data at once from template engines today.

True, but does that data need to go in one huge buffer?

**Comment by rstoyanchev** [comment]
How else would you get out of the template engine without blocking?

**Comment by poutsma** [comment]
By creating a new internal buffer every X bytes, if necessary. I believe this is what Netty and Undertow do, simply because Undertow uses buffers of 1024 bytes by default, and Netty of 256 (though auto-expanding). So let's do a reductio ad absurdum: let's assume that they wouldn't create a new buffer every X bytes. That means they wouldn't be able to serve resources larger than X bytes. And we know that they do, so therefore the assumption must be wrong.

Moreover, I would think that especially Netty would simply not accept the memory leaks described on this page. So we can assume that they don't suffer from them.

So, in conclusion: this whole "issue" falls under the "APIs don't perform if you don't use them as intended" category.

**Comment by poutsma** [comment]
Read [this comment](http://www.evanjones.ca/java-bytebuffer-leak.html#comment-2428663843) for why creating huge buffers is a bad idea.

**Comment by rstoyanchev** [comment]
Okay good to keep in mind. We can review the solution for View rendering once something is ready. Due to the blocking nature of existing template rendering libraries we have to obtain the full content (see [for example](https://github.com/ratpack/ratpack/blob/master/ratpack-groovy/src/main/java/ratpack/groovy/template/internal/MarkupTemplateRenderer.java#L57-L67)). We could split from there into smaller chunks.

**Comment by poutsma** [comment]
Or we can reintroduce the `allocateHeapBuffer` and more importantly `allocateDirectBuffer`, as it might be useful to create direct buffers for template rendering.
